graphql/src/store/prefetch.rs

```diff
@@ -506,295 +453,309 @@ fn execute_root_selection_set(
-                    Ok(children) => {
-                        let child_object_type = object_or_interface_from_type(
-                            &ctx.query.schema.document,
-                            &field.field_type,
-                        )
-                        .expect("type of child field is object or interface");
-                        match execute_selection_set(
-                            resolver,
-                            ctx,
-                            children,
-                            fields.into_iter().map(|f| &f.selection_set).collect(),
-                            &child_object_type,
-                        ) {
-                            Ok(children) => Join::perform(&mut parents, children, response_key),
-                            Err(mut e) => errors.append(&mut e),
-                        }
-                    }
-                    Err(mut e) => {
-                        errors.append(&mut e);
+            // Unwrap: The query was validated to contain only valid fields,
+            // and `collect_fields` will skip introspection fields.
+            let field = type_cond.field(&fields[0].name).unwrap();
+            let child_type = schema
+                .document
+                .object_or_interface(field.field_type.get_base_type())
+                .expect("we only collect fields that are objects or interfaces");
+
+            let join = Join::new(
+                ctx.query.schema.as_ref(),
+                type_cond,
+                child_type,
+                &field.name,
+            );
+
+            // Group fields with the same response key, so we can execute them together
+            let grouped_field_set =
+                collect_fields(ctx, child_type, fields.iter().map(|f| &f.selection_set));
+
+            match execute_field(
+                resolver, &ctx, type_cond, &parents, &join, &fields[0], field,
+            ) {
+                Ok(children) => {
+                    match execute_selection_set(resolver, ctx, children, grouped_field_set) {
+                        Ok(children) => Join::perform(parents, children, response_key),
+                        Err(mut e) => errors.append(&mut e),
                     }
-                };
-            } else {
-                errors.push(QueryExecutionError::UnknownField(
-                    fields[0].position,
-                    object_type.name().to_owned(),
-                    fields[0].name.clone(),
-                ))
-            }
+                }
+                Err(mut e) => {
+                    errors.append(&mut e);
+                }
+            };
         }
     }
 
     if errors.is_empty() {
         Ok(parents)
     } else {
-        if errors.is_empty() {
-            errors.push(QueryExecutionError::EmptySelectionSet(
-                object_type.name().to_owned(),
-            ));
-        }
         Err(errors)
     }
 }
 
+#[derive(Default)]
+struct CollectedResponseKey<'a> {
```

Comment:
Just to make sure I understand what this is (and might be good as a comment): a `CollectedResponseKey` contains a map of the `ObjectType` to the fields we fetch for that object type. If we actually fetch fields for an interface type, i.e., `iface_cond` is `Some(_)`, `iface_fields` contains the fields we fetch from the interface. Do the fields in `obj_types` already contain these fields or are they solely the type-specific fields we fetch? I'll probably find the answer to that question when I read `collect_field`, but it would be nice if it was stated explicitly.

Reply:
I've added a comment.